File: common/assert/xAssert.c
////////////////////////////////////////////////////////////
//  assert source file
//  implements assert functions
//
// general discloser: copy or share the file is forbidden
// Written : 12/01/2025
// Updated : 19/04/2025
////////////////////////////////////////////////////////////

#include "xAssert.h"
#include "xLog.h"
#include <stdlib.h>

////////////////////////////////////////////////////////////
/// xAssert
////////////////////////////////////////////////////////////
void xAssert(const uint8_t* p_ptkcFile, uint32_t p_ulLine, const char* p_ptMsg)
{
    X_LOG_ASSERT("%s", "Assertion failed");
    
    // Affiche le nom du fichier et le numéro de ligne où l'assertion a échoué
    X_LOG_ASSERT("Location: %s:%lu", p_ptkcFile, p_ulLine);

    // Affiche le message personnalisé si disponible
    if (p_ptMsg != NULL)
    {
        X_LOG_ASSERT("Message: %s", p_ptMsg);
    }

#if defined(XOS_ASSERT_MODE_EXIT)
    exit(1);
#elif defined(XOS_ASSERT_MODE_LOOP)
    while (1);
#elif defined (X_OS_RESTART_APPLICATION)
    X_LOG_ASSERT("Restarting application");

#endif
}

////////////////////////////////////////////////////////////
/// xAssertReturn
////////////////////////////////////////////////////////////
int xAssertReturn(const uint8_t* p_ptkcFile, uint32_t p_ulLine, const char* p_ptMsg, int p_iRet)
{
    X_LOG_ASSERT("Assertion failed with return value %d", p_iRet);
    
    // Affiche le nom du fichier et le numéro de ligne où l'assertion a échoué
    X_LOG_ASSERT("Location: %s:%lu", p_ptkcFile, p_ulLine);

    // Affiche le message personnalisé si disponible
    if (p_ptMsg != NULL)
    {
        X_LOG_ASSERT("Message: %s", p_ptMsg);
    }

    return p_iRet;
}

File: RUST/common/src/xAssert.rs
////////////////////////////////////////////////////////////
//  Assert header file
//  Defines assertion types and functions
//
// general discloser: copy or share the file is forbidden
// Author: Christophe
// Written: 12/01/2025
////////////////////////////////////////////////////////////


use std::process;

pub enum AssertMode 
{
    Continue,
    Exit,
    Loop,
    Restart,
}

// Configuration globale
pub static mut ASSERT_MODE: AssertMode = AssertMode::Exit;

// Macro pour les assertions
#[macro_export]
macro_rules! X_ASSERT 
{
    ($condition:expr) => {
        if !($condition) {
            $crate::xAssert(file!(), line!(), None);
        }
    };
    ($condition:expr, $message:expr) => {
        if !($condition) {
            $crate::xAssert(file!(), line!(), Some($message));
        }
    };
}

// Macro pour les assertions avec retour
#[macro_export]
macro_rules! X_ASSERT_return {
    ($condition:expr, $ret:expr) => {
        if !($condition) {
            return $crate::xAssertReturn(file!(), line!(), None, $ret);
        }
    };
    ($condition:expr, $message:expr, $ret:expr) => {
        if !($condition) {
            return $crate::xAssertReturn(file!(), line!(), Some($message), $ret);
        }
    };
}

// Fonction simplifiée de journalisation d'assertion
fn x_log_assert(message: &str) {
    eprintln!("ASSERT | {}", message);
}

pub fn xAssert(file: &str, line: u32, message: Option<&str>) {
    x_log_assert("Assertion failed");
    x_log_assert(&format!("Location: {}:{}", file, line));
    
    if let Some(msg) = message {
        x_log_assert(&format!("Message: {}", msg));
    }

    // Gestion des différents modes d'assertion
    unsafe {
        match ASSERT_MODE {
            AssertMode::Exit => {
                process::exit(1);
            },
            AssertMode::Loop => {
                loop {
                    // Boucle infinie
                    std::thread::sleep(std::time::Duration::from_secs(1));
                }
            },
            AssertMode::Restart => {
                x_log_assert("Restarting application");
                // Implémentation du redémarrage selon votre application
            },
            AssertMode::Continue => {
                // Continue l'exécution
            }
        }
    }
}

pub fn xAssertReturn<T>(file: &str, line: u32, message: Option<&str>, ret: T) -> T {
    x_log_assert("Assertion failed with return value");
    x_log_assert(&format!("Location: {}:{}", file, line));
    
    if let Some(msg) = message {
        x_log_assert(&format!("Message: {}", msg));
    }

    ret
}
